### PR TITLE
Implement backend to save course key for already created course run.

### PR DESCRIPTION
**Pull request by waheedahmed**
[ECOM-6178](https://openedx.atlassian.net/browse/ECOM-6178)

Implement the backend to save course run key in pending studio instances and send an email to course team with subject "Studio instance created".

### CODE REVIEW

**Comment by waheedahmed** [comment]
@bderusha Please review this PR, I closed previous PR https://github.com/edx/course-discovery/issues/406 and created this new one to include the ajax call from dashboard.

File: course_discovery/apps/publisher/emails.py [hunk 1]
```diff
@@ -48,4 +48,40 @@ def send_email_for_change_state(course_run):
...
+            'course_run': course_run,
+            'course_run_page_url': 'https://{host}{path}'.format(
+                host=Site.objects.get_current().domain.strip('/'), path=object_path
+            )
+        }
+
+        txt_template = 'publisher/email/studio_instance_created.txt'
+        html_template = 'publisher/email/studio_instance_created.html'
+
+        template = get_template(txt_template)
```
Comment (bderusha) [inline course_discovery/apps/publisher/emails.py]: NIT: use distinct variable names for the templates.  e.g. `txt_template = get_template(txt_template_path)`

File: course_discovery/apps/publisher/emails.py [hunk 2]
```diff
@@ -48,4 +48,40 @@ def send_email_for_change_state(course_run):
...
+        plain_content = template.render(context)
+        template = get_template(html_template)
+        html_content = template.render(context)
+        email_msg = EmailMultiAlternatives(
+            subject, plain_content, from_address, to_addresses
+        )
+        email_msg.attach_alternative(html_content, 'text/html')
+        email_msg.send()
+    except Exception:  # pylint: disable=broad-except
+        logger.exception('Failed to send email notifications for course_run [%s]', course_run.id)
```
Comment (bderusha) [inline course_discovery/apps/publisher/emails.py:85]: please add an error case in the tests for this

File: course_discovery/apps/publisher/emails.py [hunk 3]
```diff
@@ -48,4 +48,40 @@ def send_email_for_change_state(course_run):
...
+    """ Send an email to course team on studio instance creation.
+
+        Arguments:
+            course_run (CourseRun): CourseRun object
+    """
+    try:
+        object_path = reverse('publisher:publisher_course_run_detail', kwargs={'pk': course_run.id})
+        subject = _('Studio instance created')
+
+        to_addresses = course_run.course.get_group_users_emails()
```
Comment (bderusha) [inline course_discovery/apps/publisher/emails.py:64]: Question:  Can we BCC these email addresses?

File: course_discovery/static/js/publisher/views/dashboard.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+            url: courseRunPageURL,
+            type: "PATCH",
+            data: JSON.stringify({lms_course_id: courseKeyValue}),
+            contentType: "application/json",
+            headers: headers,
+            success: function (response) {
+                $courseRunParentTag.remove();
+                $studioInstanceSuccess.find(".copy-meta").html(successMessage);
+                $studioInstanceSuccess.css("display", "block");
+                alertTimeout(10000, '.studio-instance-success');
```
Comment (bderusha) [inline course_discovery/static/js/publisher/views/dashboard.js]: Is this part of any acceptance criteria?  I don't remember seeing anything about the message going away.

File: course_discovery/static/js/publisher/views/dashboard.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            courseTitleTag = $courseRunParentTag.find("#course-title").html(),
+            startDateTag = $courseRunParentTag.find("#course-start").html(),
+            headers = {
+                'X-CSRFToken': Cookies.get('course_discovery_csrftoken')
+            },
+            $studioInstanceSuccess = $(".studio-instance-success"),
+            successMessage = interpolateString(
+                gettext("You have successfully created a studio instance ({studioLinkTag}) for {courseRunDetail} with a start date of {startDate}"),
+                {
+                    "studioLinkTag": "<a hrep=''>"+ courseKeyValue +"</a>",
```
Comment (bderusha) [inline course_discovery/static/js/publisher/views/dashboard.js]: `href`

**Review by bderusha — CHANGES_REQUESTED** [review]

File: course_discovery/apps/publisher/emails.py (hunk 3 above)
Comment (awais786) [reply course_discovery/apps/publisher/emails.py:64]: @waheedahmed add a separate story for this.

Comment (waheedahmed) [reply course_discovery/apps/publisher/emails.py:64]: @bderusha Sure, done.

@awais786 I think you can create a new story for other email functionalities.

**Comment by waheedahmed** [comment]
Dear @bderusha addressed all your comments, please take another look.

File: course_discovery/apps/publisher/emails.py [hunk 6]
```diff
@@ -48,4 +48,38 @@ def send_email_for_change_state(course_run):
...
```
Comment (awais786) [inline course_discovery/apps/publisher/emails.py:64]: bcc_addresses

File: course_discovery/apps/publisher/views.py [hunk 7]
```diff
@@ -287,3 +292,8 @@ def post(self, request):
         user_attribute.save()
 
         return JsonResponse({'is_enabled': is_enabled})
+
+
+class UpdateCourseKeyView(mixins.LoginRequiredMixin, mixins.ViewPermissionMixin, UpdateAPIView):
```
Comment (awais786) [inline course_discovery/apps/publisher/views.py:276]: Why using ViewPermissionMixin here ?

File: course_discovery/apps/publisher/emails.py [hunk 8]
```diff
@@ -71,16 +71,14 @@ def send_email_for_studio_instance_created(course_run):
...
-            subject, plain_content, from_address, to_addresses
-        )
+        txt_template_path = 'publisher/email/studio_instance_created.txt'
+        html_template_path = 'publisher/email/studio_instance_created.html'
+
+        txt_template = get_template(txt_template_path)
+        plain_content = txt_template.render(context)
+        html_template = get_template(html_template_path)
+        html_content = html_template.render(context)
+        email_msg = EmailMultiAlternatives(subject, plain_content, from_address, bcc=to_addresses)
```
Comment (bderusha) [inline course_discovery/apps/publisher/emails.py]: Just want to double check that this works without anything in the `to` field.  If not we may need to setup some dummy `to` address.  Can you confirm either way?

**Review by bderusha — APPROVED** [review]

File: course_discovery/apps/publisher/emails.py (hunk 8 above)
Comment (waheedahmed) [reply course_discovery/apps/publisher/emails.py]: @bderusha Its working fine with nothing in `to` field.

Comment (waheedahmed) [reply course_discovery/apps/publisher/emails.py]: screen shot 2016-11-04 at 2 51 39 pm

**Review by awais786 — APPROVED** [review]

File: course_discovery/apps/publisher/views.py (hunk 7 above)
Comment (waheedahmed) [reply course_discovery/apps/publisher/views.py:276]: @awais786 We need to check the permission of the user on object who is adding the course key.

File: course_discovery/apps/publisher/emails.py (hunk 6 above)
Comment (waheedahmed) [reply course_discovery/apps/publisher/emails.py:64]: I think no need to change this.

**Comment by waheedahmed** [comment]
I'll merge it when @awais786's PR https://github.com/edx/course-discovery/pull/421 is ready to merge.

**Comment by awais786** [comment]
@waheedahmed @bderusha I am closing this PR. I have opened the new PR.
I have already merged my PR into master and this PR is created from my branch. So I have to close it create the new one from master.
New PR https://github.com/edx/course-discovery/pull/435

**Comment by awais786** [comment]
Also bcc is not working with django-ses. So I have moved this back to `to`.  I will file a separate to implement this. It is working with SMTP emails but not with django-ses.

**Pull request closed without merging**